app/models/provider_user.rb

```diff
@@ -12,7 +12,8 @@ class ProviderUser < ActiveRecord::Base
 
   scope :visible_to, ->(provider_user) {
     joins(:provider_permissions)
-      .where(ProviderPermissions.table_name => { provider_id: provider_user.providers.pluck(:id) })
+      .where(ProviderPermissions.table_name => { provider_id: provider_user.provider_permissions.where(manage_users: true).pluck(:provider_id) })
```

Comment:
I don't really grasp how this will return duplicate ProviderUsers in the first place? A normal `Model.where()` could never do that, or am I confused? 

Does it have something to do with the `joins`? Does it need to be a left/right/inner join instead of whatever it is now? 

Or does it have something to do with the fact that we do 2 queries when calling this scope? Might be worth doing this so it becomes a subquery:

```suggestion
      .where(ProviderPermissions.table_name => { provider_id: provider_user.provider_permissions.where(manage_users: true).select(:provider_id) })
```

Reply:
Good shout on the subquery — will add. The generated SQL looks like this:

```
SELECT "provider_users".* FROM "provider_users" INNER JOIN "provider_users_providers" ON "provider_users_providers"."provider_user_id
" = "provider_users"."id" WHERE "provider_users_providers"."provider_id" IN (SELECT "providers"."id" FROM "providers" INNER JOIN "provider_users_providers" ON
 "providers"."id" = "provider_users_providers"."provider_id" WHERE "provider_users_providers"."provider_user_id" = 5);
```

Or with the subquery evaluated:

```
 SELECT "provider_users".* FROM "provider_users" INNER JOIN "provider_users_providers" ON "provider_users_providers"."provider_user_id" = "provider_users"."id" WHERE "provider_users_providers"."provider_id" IN (5, 1, 6)
```

In either case, you can see the dupes racking up:

![Screenshot 2020-04-21 at 16 11 04](https://user-images.githubusercontent.com/642279/79882573-bd50da80-83ea-11ea-9d7b-f2285d4f82bb.png)

I guess the INNER JOIN might be distinctly satisfied by each provider_user_id/provider_id combination, as they're part of the where? 

I initially solved this with `GROUP BY` provider_users.id but preferred `DISTINCT`.

Reply:
Ok, I tried to solve this with `left_joins` but I got really confused!

Here's a single-query scope that does the same thing, in case you find it simpler to read (I'm not sure it's more performant):

```rb
  scope :visible_to, ->(provider_user) {
    providers_that_user_can_manage = provider_user.provider_permissions.manage_users.select(:provider_id)

    users_that_user_can_see = ProviderPermissions.where(
      provider_id: providers_that_user_can_manage,
    ).select(:provider_user_id)

    where(id: users_that_user_can_see)
  }

```

Reply:
More readable, it's going in!